### PR TITLE
Add CircuitBreaker to TDigest, Step 4: Take into account shallow classes size

### LINKED ISSUE #113916: [CI] EsqlActionBreakerIT class failing
**Build Scans:**
- [elasticsearch-periodic-platform-support #4240 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/fi5d462bc7jl6)
- [elasticsearch-pull-request #33610 / part-3](https://gradle-enterprise.elastic.co/s/fu4mxturtsae2)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:internalClusterTest" --tests "org.elasticsearch.xpack.esql.action.EsqlActionBreakerIT.testFromGroupingByNumericFieldWithNulls" -Dtests.seed=87843C93CF72D74C -Dtests.locale=ff-GN -Dtests.timezone=Etc/GMT+4 -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.action.EsqlActionBreakerIT'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.action.EsqlActionBreakerIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: Request breaker not reset to 0 on node: node_s1
Expected: <0L>
     but: was <48L>
```

**Issue Reasons:**
- [main] 2 failures in class org.elasticsearch.xpack.esql.action.EsqlActionBreakerIT (0.6% fail rate in 311 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### PR DESCRIPTION
**Pull request by ivancea**
Final (I wish) part of https://github.com/elastic/elasticsearch/issues/99815
Also, fixes https://github.com/elastic/elasticsearch/issues/113916

## Steps
1. Migrate TDigest classes to use a custom Array implementation. Temporarily use a simple array wrapper (https://github.com/elastic/elasticsearch/pull/112810)
2. Implement CircuitBreaking in the `WrapperTDigestArrays` class. Add Releasable/AutoCloseable and ensure everything is closed (https://github.com/elastic/elasticsearch/pull/113105)
3. Pass the CircuitBreaker as a parameter to TDigestState from wherever it's being used (https://github.com/elastic/elasticsearch/pull/113387)
    - ESQL: Pass a real CB
    - Other aggs: Use the deprecated methods on `TDigestState`, that will use a No-op CB instead
4. Account remaining TDigest classes size ("SHALLOW_SIZE") (This PR)

Every step should be safely mergeable to main:
- The first and second steps should have no impact.
- The third and fourth ones will start increasing the CB count partially.

## Remarks
As TDigests are releasable now, I had to refactor all tests, adding try-with-resources or direct close() calls. That added a lot of changes, but most of them are trivial.

Outside of it, in ESQL, TDigestStates are closed now.
Old aggregations don't close them, as it's not trivial. However, as they are using the NoopCircuitBreaker, there's no problem with it. There's nothing to be closed.

## _Remarks 2_
I tried to follow the same pattern in how everything is accounted. On each TDigest class:
- Static constant "SHALLOW_SIZE" with the object weight
- Field `AtomicBoolean closed` to ensure indempotent `close()`
- Static `create()` method that accounts the SHALLOW_SIZE, and returns a new isntance. And the important part: On exception, it discounts SHALLOW_SIZE again
- A `ramBytesUsed()` (Accountable interface), barely used for anything really, but some assertions I believe
- A constructor, that closes everything it created on exception (If it creates an array, and the next array surpasses the CB limit, the first one must be closed)
- And a close() that will, well, close everything and discount SHALLOW_SIZE

A lot of steps to make sure everything works well in this multi-level structure, but I believe the result was quite clean

### CODE REVIEW

File: libs/tdigest/build.gradle [hunk 1]
```diff
@@ -23,6 +23,7 @@ apply plugin: 'elasticsearch.publish'
 
 dependencies {
   api project(':libs:elasticsearch-core')
+  api "org.apache.lucene:lucene-core:${versions.lucene}"
```
Comment (ivancea) [inline libs/tdigest/build.gradle:26]: For `Accountable` and `RamUsageEstimator`

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java [hunk 2]
```diff
@@ -42,49 +50,95 @@ final class AVLGroupTree extends AbstractCollection<Centroid> implements Releasa
...
+            throw e;
+        }
+    }
 
-            @Override
-            protected void copy(int node) {
-                centroids.set(node, centroid);
-                counts.set(node, count);
-            }
+    private class InternalIntAvlTree extends IntAVLTree {
```
Comment (ivancea) [inline libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java:96]: Note that this isn't a static class, as I just moved the anonymous class from the constructor. So this class doesn't have a `static create()` method, and it's in the parent class instead

File: libs/tdigest/build.gradle (hunk 1 above)
Comment (nik9000) [reply libs/tdigest/build.gradle:26]: :+1:

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java [hunk 3]
```diff
@@ -29,11 +31,17 @@
...
 
 /**
  * A tree of t-digest centroids.
  */
-final class AVLGroupTree extends AbstractCollection<Centroid> implements Releasable {
+final class AVLGroupTree extends AbstractCollection<Centroid> implements Releasable, Accountable {
+    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(AVLGroupTree.class);
+
+    private final TDigestArrays arrays;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
```
Comment (nik9000) [inline libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java:43]: Are these thread safe classes other than this? If not I'd prefer we not use `Atomic` stuff. *most* of aggs isn't thread safe and doesn't have to be intentionally.

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/HybridDigest.java [hunk 4]
```diff
@@ -49,18 +53,28 @@ public class HybridDigest extends AbstractTDigest {
...
+
     /**
      * Creates a hybrid digest that uses a {@link SortingDigest} for up to {@param maxSortingSize} samples,
      * then switches to a {@link MergingDigest}.
      *
      * @param compression The compression factor for the MergingDigest
      * @param maxSortingSize The sample size limit for switching from a {@link SortingDigest} to a {@link MergingDigest} implementation
      */
-    HybridDigest(TDigestArrays arrays, double compression, long maxSortingSize) {
+    private HybridDigest(TDigestArrays arrays, double compression, long maxSortingSize) {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/HybridDigest.java:73]: :+1:

File: libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java (hunk 3 above)
Comment (nik9000) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java:43]: I'm not worried about the overhead, not really. I'm more worried that a reader will see this and say "Oh! this is shared in many threads" and spend some time trying to find that when it isn't really shared.

Comment (ivancea) [reply libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLGroupTree.java:43]: I used AtomicBoolean to replicate the pattern we use along ES (From what I saw, most other classes that check for close() idempotence do that).
But no, those classes aren't thread-safe afaik. So I can change it if it feels better

**Review by nik9000 — APPROVED** [review]

**Pull request merged**